core/server/common/src/main/java/alluxio/master/journal/JournalCheckpointThread.java

```diff
@@ -0,0 +1,217 @@
+ * the checkpoint being written will be cancelled.
+ */
+@NotThreadSafe
+public final class JournalCheckpointThread extends Thread {
+  private static final Logger LOG = LoggerFactory.getLogger(JournalCheckpointThread.class);
+
+  /** The master to apply the journal entries to. */
+  private final Master mMaster;
+  /** The journal. */
+  private final Journal mJournal;
+  /** Make sure no new journal logs are found for this amount of time before shutting down. */
+  private final int mShutdownQuietWaitTimeMs;
+  /** If not journal log is found, sleep for this amount of time and check again. */
+  private final int mJournalCheckpointSleepTimeMs;
+  /** This becomes true when the master initiates the shutdown. */
+  private volatile boolean mInitiateShutdown = false;
+
+  /** True if this thread is no longer running. */
+  private volatile boolean mStopped = false;
+
+  /** The journal reader. */
+  private JournalReader mJournalReader;
+
+  /**
+   * Creates a new instance of {@link JournalCheckpointThread}.
+   *
+   * @param master the master to apply the journal entries to
+   * @param journal the journal
+   */
+  public JournalCheckpointThread(Master master, Journal journal) {
+    mMaster = Preconditions.checkNotNull(master);
+    mJournal = Preconditions.checkNotNull(journal);
+    mShutdownQuietWaitTimeMs =
+        Configuration.getInt(PropertyKey.MASTER_JOURNAL_TAILER_SHUTDOWN_QUIET_WAIT_TIME_MS);
+    mJournalCheckpointSleepTimeMs =
+        Configuration.getInt(PropertyKey.MASTER_JOURNAL_TAILER_SLEEP_TIME_MS);
+    mJournalReader = mJournal.getReader(JournalReaderCreateOptions.defaults().setPrimary(false));
+  }
+
+  /**
+   * Initiates the shutdown of this checkpointer thread, and also waits for it to finish.
+   */
+  public void shutdownAndJoin() {
+    LOG.info("{}: Journal checkpointer shutdown has been initiated.", mMaster.getName());
+    mInitiateShutdown = true;
+
+    try {
+      // Wait for the thread to finish.
+      join();
+    } catch (InterruptedException e) {
+      LOG.error("{}: journal checkpointer shutdown is interrupted.", mMaster.getName(), e);
+      // Kills the master. This can happen in the following two scenarios:
+      // 1. The user Ctrl-C the server.
+      // 2. Zookeeper selects this master as standby before the master finishes the previous
+      //    standby->leader transition. It is safer to crash the server because the behavior is
+      //    undefined to have two journal checkpointer running concurrently.
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * @return the last edit log sequence number read plus 1
+   */
+  public long getNextSequenceNumber() {
+    Preconditions.checkState(mStopped);
+    return mJournalReader.getNextSequenceNumber();
+  }
+
+  @Override
+  public void run() {
+    try {
+      runInternal();
+    } catch (RuntimeException e) {
+      LOG.error("{}: Failed to run journal checkpoint thread, crashing.", mMaster.getName(), e);
+      System.exit(-1);
```

Comment:
Does it matter if it exits? I guess not since all the resources should be gone after the process is dead. I've moved this to runInternal().  The behavior is the same (not closed when runtime exception is thrown).